contributors/design-proposals/service-injection-policy.md

```diff
@@ -0,0 +1,444 @@
+    Policy is added that would effect them.
+
+## Use Cases
+
+1.  As a user, I want to be able to describe a way that pods should be injected
+    with the information to consume a particular service. This would allow me
+    to concisely model the information about how the service
+    should be consumed without altering every consuming pod spec.
+
+### Loose coupling between services and their consumers
+
+Using a Service Injection Policy allows pod template authors to not have to explicitly
+set information for every pod. This way authors of pod templates consuming a
+specific service do not need to know all the details about that service.
+
+The information will be injected based off the policy's `Selector`.
+
+## Proposed Changes
+
+### ServiceInjectionPolicy API object
+
+This resource is alpha. The policy itself is immutable.
+
+```go
+type ServiceInjectionPolicy struct {
+    unversioned.TypeMeta
+    ObjectMeta
+
+    Spec ServiceInjectionPolicySpec
+}
+
+type ServiceInjectionPolicySpec struct {
+    // Selector defines which Pods should be modified, this only targets Pods.
+    Selector      unversioned.LabelSelector
+    // Env defines the collection of EnvVar to inject.
+    Env           []EnvVar `json:omitempty`
+    // EnvFrom defines the collection of EnvFromSource to inject.
+    EnvFrom       []EnvFromSource `json:omitempty`
+    // Volumes defines the collection of Volume to inject.
+    Volumes       []Volume `json:omitempty`
+    // VolumeMounts defines the collection of VolumeMount to inject.
+    VolumeMounts  []VolumeMount `json:omitempty`
+}
+```
+
+#### Validations
+
+In order for the Service Injection Policy to be valid it must fulfill the
+following constraints:
+
+- The `Selector` field must be defined. This is how we know which pod templates
+  to inject so therefore it is required.
+- The policy must define _at least_ 1 of `Env`, `EnvFrom`, `Volumes`, or
+  `VolumeMounts`.
+- The `Volume`s given in `Volumes` must exist and be valid.
+
+#### Conflicts
+
+There are a number of edge conditions that might occur at the time of
+injection. These are as follows:
+
+- Merging lists with no conflicts: if a pod already has a volume defined
+  exactly as defined in the SIP. This would replace the other volume.
+- Merging lists with conflicts: if a SIP redefines an `EnvVar` or a `Volume`,
+  the original will be replaced and an annotation to the Pod spec will reflect
+  a warning documenting the decision.
+- Conflicts between `Env` and `EnvFrom`: this would throw an error.
+
+### AdmissionControl Plug-in: ServiceInjectionPolicy
+
+The **ServiceInjectionPolicy** plug-in introspects all incoming pod creation
+requests and injects the pod based off a `Selector` with the desired
+attributes.
+
+For the initial alpha, the order of precedence for applying multiple
+`ServiceInjectionPolicy` specs is from oldest to newest. All Service Injection
+Policies in a namespace should be order agnostic. However we can use merge keys
```

Comment:
Are these two sentence inconsistent?  Meaning, we say order matters (newest wins) but then we say they should be order agnostic in a namespace. We should pick one rule and stick to it. And I'm not sure what "agnostic" means in this case - if we mean the results are "indeterminate" then I think it would be better to say so.